# Demonstração do código vetorizado para o sistema de 14 barras


# Importando as bibliotecas necessárias

In [1]:
import numpy as np
import timeit
import copy

# Funções auxiliares

Lista de funções:

* discreto_superior(vetor\_x, lista\_discretos)

    Exemplo 1:
      Entrada:
        vetor_x = [0.0, 0.15, 0.30, 0.39]
        lista_discretos = [0.0, 0.19, 0.34, 0.39]

      Saída:
        x_sup = [0.19, 0.19, 0.34, 0.39]

    Exemplo 2:
       Entrada:
        vetor_x = [1.0, 1.5, 2.1, 3.0, 4.0, 4.5]
        lista_discretos = [0.0, 2.0, 3.5, 4.5] 
        
      Saída:
        x_sup = [2.0, 2.0, 3.5, 3.5, 4.5, 4.5]

* discreto_inferior(vetor\_x, lista\_discretos)

    Exemplo 1:
      Entrada:
        vetor_x = [0.0, 0.15, 0.30, 0.39]
        lista_discretos = [0.0, 0.19, 0.34, 0.39]

      Saída:
        x_sup = [0.0, 0.0, 0.19, 0.34]

    Exemplo 2:
       Entrada:
        vetor_x = [1.0, 1.5, 2.1, 3.0, 4.0, 4.5]
        lista_discretos = [0.0, 2.0, 3.5, 4.5] 
        
      Saída:
        x_sup = [0.0, 0.0, 2.0, 2.0, 3.5, 3.5]


In [2]:
def discreto_superior(vetor_x, lista_discretos):
    '''
    Função que retorna o valor discreto superior de 'lista_discretos' mais próximo de todos os valores x de 'vetor_x' 

    Inputs:
        -> vetor_x = vetor (numpy array) contendo os valores que deseja-se obter o número discreto mais próximos
        -> lista_discretos = lista (python list) que contém o conjunto de valores discretos que cada variável x admite
    
    Ouputs:
        -> x_sup = vetor (numpy array) contendo os valores discretos superiores de 'lista_discretos' mais próximo 
           dos valores de 'vetor_x'
    '''
    #Vetor de saída da função. Possui o mesmo formato (shape) que vetor_x
    x_sup = np.zeros(vetor_x.shape)
    
    #Cópia de 'vetor_x'. Esta cópia é feita para evitar erros de alocamento dinâmico de memória.
    vetor = np.copy(vetor_x)
    
    #GAMBIARRA SEM A QUAL O CÓDIGO NÃO FUNCIONA PROS VALORES DE 'vetor_x' que pertençam a 'lista_discretos'
    #vetor = vetor + 1e-3
    
    #Garante que a lista seja uma array numpy e armazena o resultado na variável 'lista'
    lista = np.asarray(lista_discretos, dtype = np.float32)
    
    '''
    Garante que os valores de 'vetor_x' estejam dentro dos limites de 'lista discretos' por um pequeno fator de 10^-3.
    Caso contrário, a função numpy.searchsorted descrita a frente resultará em erro.
    '''
    np.clip(a = vetor, a_min = lista[0]+1e-3, a_max = lista[-1]-1e-3, out = vetor)
    
    '''
    Utilizando a função numpy.searchsorted() para buscar os índices de 'lista_discretos' que correspondem aos valores
    discretos superiores aos valores de 'vetor_x'
    '''
    indices = np.searchsorted(a=lista, v = vetor, side='right')
    
    #Armazena os valores de 'lista_discretos' cujos índices correspondem aos discretos superiores de 'vetor_x'
    x_sup = np.take(lista, indices)
    
    #Deleta as variáveis locais
    del vetor, lista, indices
    
    return x_sup

In [3]:
'''
Teste de discreto_superior
'''
vetor_teste = np.array([0., 0.15, 0.19, 0.30, 0.34, 0.37, 0.39])
lista_teste = np.array([0, 0.19, 0.34, 0.39])

print('Vetor_teste = {}'.format(vetor_teste))
print('x_sup = {}'.format(discreto_superior(vetor_teste, lista_teste)))

Vetor_teste = [0.   0.15 0.19 0.3  0.34 0.37 0.39]
x_sup = [0.19 0.19 0.34 0.34 0.34 0.39 0.39]


Resultado esperado:

    x_sup = [0.19, 0.19, 0.34, 0.34, 0.39, 0.39, 0.39]


In [4]:
def discreto_inferior(vetor_x, lista_discretos):
    '''
    Função que retorna o valor discreto inferior de 'lista_discretos' mais próximo de todos os valores x de 'vetor_x' 
    
    Inputs:
        -> vetor_x = vetor (numpy array) contendo os valores que deseja-se obter o número discreto mais próximo
        -> lista_discretos = lista (python list) que contém o conjunto de valores discretos que cada variável x admite
    
    Ouputs:
        -> x_inf = vetor (numpy array) contendo os valores discretos inferiores de 'lista_discretos' mais próximos 
           dos valores de 'vetor_x'
    '''
    
    #Vetor de saída da função. Possui o mesmo formato (shape) que vetor_x
    x_inf = np.zeros(vetor_x.shape)
    
    #Cópia de 'vetor_x'. Esta cópia é feita para evitar erros de alocamento dinâmico de memória.
    vetor = np.copy(vetor_x)
    
    #GAMBIARRA SEM A QUAL O CÓDIGO NÃO FUNCIONA PROS VALORES DE 'vetor_x' que pertençam a 'lista_discretos'
    #vetor = vetor - 1e-3
    
    #Garante que a lista seja uma array numpy e salva o resultado na variável local 'lista'
    lista = np.asarray(lista_discretos, dtype = np.float32)
    
    '''
    Garante que os valores de 'vetor_x' estejam dentro dos limites de 'lista discretos' por um pequeno fator de 10^-3.
    Caso contrário, a função numpy.searchsorted descrita a frente resultará em erro. Salva o resultado de numpy.clip
    na variável local 'vetor'
    '''
    np.clip(a = vetor, a_min = lista_discretos[0]+1e-3, a_max = lista_discretos[-1]-1e-3, out = vetor)
    '''[np.nonzero(vetor)]'''
    
    
    '''
    Utilizando a função numpy.searchsorted() para buscar os índices de 'lista_discretos' que correspondem aos valores
    discretos inferiores aos valores de 'vetor_x'
    '''
    indices = np.searchsorted(a=lista, v = vetor, side='left') - 1
    
    #Armazena os valores de 'lista_discretos' cujos índices correspondem aos discretos superiores de 'vetor_x'
    x_inf = np.take(lista, indices)
    
    #Deleta as variáveis locais
    del vetor, lista, indices
    
    return x_inf

In [5]:
'''
Teste de discreto_inferior
'''
vetor_teste = np.array([0, 0.15, 0.19, 0.30, 0.34, 0.37, 0.39])
lista_teste = [0, 0.19, 0.34, 0.39]

print('Vetor_teste = {}'.format(vetor_teste))
print('x_inf = {}'.format(discreto_inferior(vetor_teste, lista_teste)))

Vetor_teste = [0.   0.15 0.19 0.3  0.34 0.37 0.39]
x_inf = [0.   0.   0.19 0.19 0.19 0.34 0.34]


Resultado esperado:

    x_inf = [0.0, 0.0, 0.0, 0.19, 0.19, 0,34, 0.34]

# Gerenciamento da rede

In [6]:
import pandapower as pp
from pandapower.networks import case14

rede = case14()

In [7]:
def gerenciar_rede(rede):
    
    """
    Esta funcao organiza a rede obtida do PandaPower de modo que ela possa ser mais facilmente utilizada pelo algoritmo.
    Suas funcionalidades são:
        
        -> Ordenar os parâmetros da rede (v_bus, tap, shunt, etc) por índices;
        -> Obter os transformadores com controle de tap;
        -> Gerar o vetor com os valores dos taps dos transformadores;
        -> Gerar os vetores com os valores discretos para os shunts de cada sistema;
        -> Gerar o primeiro agente de busca (que contém as variáveis do ponto de operação do sistema);
        -> Obter as condutâncias das linhas.
        
    Input:
        -> rede
        
    Output:
        
        -> rede gerenciada (não devolvida, salva diretamente na variável rede);
        -> primeiro agente de buscas: lobo_1;
        -> vetor de condutâncias da rede: G_rede;
        -> matriz das linhas de transmissao: linhas
    """
    
    #Ordenar os índices da rede
    rede.bus = rede.bus.sort_index()
    rede.res_bus = rede.res_bus.sort_index()
    rede.gen = rede.gen.sort_index()
    rede.line = rede.line.sort_index()
    rede.shunt = rede.shunt.sort_index()
    rede.trafo = rede.trafo.sort_index()
    
    
    
    #num_trafo_controlado: variavel para armazenar o número de trafos com controle de tap
    num_trafo_controlado = rede.trafo.tap_pos.count()
    
    #num_barras : variavel utilizada para salvar o numero de barras do sistema
    num_barras = rede.bus.name.count()
    
    #num_shunt: variavel para armazenar o numero de shunts do sistema
    num_shunt = rede.shunt.in_service.count()
    
    #num_gen: variavel para armazenar o número de barras geradoras do sistema
    num_gen = rede.gen.in_service.count()
    
    '''
    Cria as varíaveis globais nb, nt, ns, ng para facilitar o uso desses parâmetros em outros funções
    
    *Potencialmente desastroso*
    '''
    global nb, nt, ns, ng
    nb, nt, ns, ng = num_barras, num_trafo_controlado, num_shunt, num_gen
    
    '''
    Muda os valores máximos e mínimos permitidos das tensões das barras dos sistemas de 118 e 300 barras:
        min_vm_pu: 0.94 -> 0.90
        max_vm_pu: 1.06 -> 1.10
    '''
    if num_barras == 118 or num_barras == 300:
        rede.bus.min_vm_pu = 0.90
        rede.bus.max_vm_pu = 1.10
    
    #Dicionário que contem os valores dos shunts para cada sistema IEEE
    valores_shunts = {"14": [[0, 0.19, 0.34, 0.39]],
                      "30": [[0, 0.19, 0.34, 0.39],
                             [0, 0.05, 0.09]],
                      "57": [[0, 0.12, 0.22, 0.27], 
                             [0, 0.04, 0.07, 0.09], 
                             [0, 0.1, 0.165]],
                      "118": [[-0.4, 0],
                              [0, 0.06, 0.07, 0.13, 0.14, 0.2],
                              [-0.25, 0],
                              [0, 0.1],
                              [0, 0.1],
                              [0, 0.1],
                              [0, 0.15],
                              [0, 0.08, 0.12, 0.2],
                              [0, 0.1, 0.2],
                              [0, 0.1, 0.2],
                              [0, 0.1, 0.2],
                              [0, 0.1, 0.2],
                              [0, 0.06, 0.07, 0.13, 0.14, 0.2],
                              [0, 0.06, 0.07, 0.13, 0.14, 0.2]],
                      "300": [[0, 2, 3.5, 4.5],
                              [0, 0.25, 0.44, 0.59],
                              [0, 0.19, 0.34, 0.39],
                              [-4.5, 0],
                              [-4.5, 0],
                              [0, 0.25, 0.44, 0.59],
                              [0, 0.25, 0.44, 0.59],
                              [-2.5, 0],
                              [-4.5, 0],
                              [-4.5, 0],
                              [-1.5, 0],
                              [0, 0.25, 0.44, 0.59],
                              [0, 0,15],
                              [0, 0.15]]
                      }
    
    #Vetor que contém os valores discretos dos taps: entre 0.9 e 1.1 com passo = tap_step
    #Precisa ser um tensor de rank 1 para que a alcateia possa ser inicializada
    global tap_step
    tap_step = 0.00625
    valores_taps = np.arange(start = 0.9, stop = 1.1, step = tap_step)
    
    """
    Matriz contendo as linhas de transmissão da rede:
        -> linhas[0] = vetor com as barras de ínicio;
        -> linhas[1] = vetor com as barras de términa;
        -> linhas[2] = vetor com as resistências em pu (r_pu) das linhas;
        -> linhas[3] = vetor com as reatâncias em pu (x_pu) das linhas.
    r_pu = r_ohm/z_base
    x_pu = x_ohm/z_base
    g = r_pu/(r_pu^2 + x_pu^2)
    """
    
    linhas = np.zeros((4, rede.line.index[-1]+1))
    linhas[0] = rede.line.from_bus.to_numpy()
    linhas[1] = rede.line.to_bus.to_numpy()
    v_temp = rede.bus.vn_kv.to_numpy()
    z_base = np.power(np.multiply(v_temp,1000), 2)/100e6
    for i in range(rede.line.index[-1]+1):
        linhas[2][i] = rede.line.r_ohm_per_km[i]/z_base[int(linhas[0][i])]
        linhas[3][i] = rede.line.x_ohm_per_km[i]/z_base[int(linhas[0][i])]
    del v_temp, z_base
    
    #Vetor G_rede com as condutâncias das linhas de transmissão
    G_rede = np.zeros((1, rede.line.index[-1]+1))
    G_rede = np.array([np.divide(linhas[2], np.power(linhas[2],2)+np.power(linhas[3],2))])
    
    #Matriz de condutância nodal da rede. É equivalente à parte real da matriz de admintância nodal do sistema
    matriz_G = np.zeros((num_barras,num_barras))
    matriz_G[linhas[0].astype(np.int), linhas[1].astype(np.int)] = G_rede 
    
    """
    O primeiro lobo (agente de busca) será inicializado com os valores de operação da rede fornecidos
    pelo PandaPower: vetor lobo_1.
    
    tap_pu = (tap_pos + tap_neutral)*tap_step_percent/100 (equação fornecida pelo PandaPower)
    shunt_pu = -100*shunt (equação fornecida pelo PandaPower)
    
    As variáveis v_temp, taps_temp e shunt_temp são utilizadas para receber os valores de tensão, tap e shunt da rede
    e armazenar no vetor lobo_1
    """
    
    v_temp = rede.gen.vm_pu.to_numpy(dtype = 'float64')
    
    taps_temp = 1 + ((rede.trafo.tap_pos.to_numpy()[0:num_trafo_controlado] +\
                      rede.trafo.tap_neutral.to_numpy()[0:num_trafo_controlado]) *\
                     (rede.trafo.tap_step_percent.to_numpy()[0:num_trafo_controlado]/100))
        
    shunt_temp = -rede.shunt.q_mvar.to_numpy()/100
    
    lobo_1 = np.array([np.concatenate((v_temp, taps_temp, shunt_temp),axis=0)])
    del v_temp, taps_temp, shunt_temp
      
    parametros_rede = {"Linhas": linhas,
                       "G": G_rede,
                       "Lobo1": lobo_1,
                       "Valores_shunts": valores_shunts,
                       "Valores_taps": valores_taps,
                       "num_barras": num_barras,
                       "num_trafo_controlado": num_trafo_controlado,
                       "num_shunt": num_shunt,
                       "num_gen": num_gen,
                       "matriz_G": matriz_G}

    return parametros_rede

In [8]:
'''
Teste de gerenciar_rede()
'''

parametros_rede = gerenciar_rede(rede)

conjunto_shunts = parametros_rede["Valores_shunts"][str(nb)]
print('O conjunto de valores discretos permitido para o shunt é: {}'.format(conjunto_shunts))

O conjunto de valores discretos permitido para o shunt é: [[0, 0.19, 0.34, 0.39]]


Resultado esperado:

    Conjunto_shunts = [[0.0, 0.19, 0.34, 0.39]]

# Inicializando a alcateia

In [9]:
def inicializar_alcateia(n_lobos, rede, parametros_rede):
    '''
    Esta função inicializa a alcateia de lobos (agentes de busca) como uma matriz (numpy array) com formato (dim+6, n_lobos)
    
        -> Cada coluna da matriz representa um lobo.
        -> As linhas de 0 a dim são as variáveis do problema.
        -> A linha dim+1 armazena a função fitness para cada lobo;
        -> A linha dim+2 armazena a função objetivo para cada lobo;
        -> A linha dim+3 armazena um caracter que indica a posição hierárquica do lobo:
            + 1 = lobo alfa;
            + 2 = lobo beta;
            + 3 = lobo delta;
            + 4 = lobo omega.
        -> A linha dim+4 armazena a penalidade dos taps para cada lobo;
        -> A linha dim+5 armazena a penalidade dos shunts para cada lobo;
        -> A linha dim+6 armazena a penalidade das tensões para cada lobo;
        
    Inputs:
        -> n_lobos = número de agentes de busca;
        -> parametros_rede - obtido via função gerenciar_rede
    
    Output:
        -> Alcateia
    '''
    
    #Variável que armazena o número de variáveis do problema
    dim = ng+nt+ns
    
    #Inicialização da Alcateia com zeros
    alcateia = np.zeros((dim+6, n_lobos))
    
    #Inicialização aleatória das variáveis contínuas (tensões das barras geradoras) a partir de uma distribuição normal
    alcateia[:ng, :] = np.random.uniform(rede.bus.min_vm_pu[0], rede.bus.max_vm_pu[1], size=(ng,n_lobos))
    
    #Inicialização dos taps dos transformadores a partir da escolha aleatória dentro dos valores discretos permitidos
    alcateia[ng:ng+nt, :] = np.random.choice(parametros_rede["Valores_taps"], size =(nt, n_lobos))
    
    #Inicialização dos shunts das barras a partir da escolha aleatória dentro dos valores discretos permitidos
    #Não consegui escapar do loop de for aqui ainda =/
    for i in range(ns):
        alcateia[ng+nt+i, :] = np.random.choice(parametros_rede["Valores_shunts"][str(nb)][i], size = (ns,n_lobos))
    
    #Inicializar a função fitness e a função objetivo de cada lobo
    alcateia[dim:dim+2, :] = np.inf
    
    #Inicializar a posição de cada lobo como delta
    alcateia[dim+2, :] = 4
    
    #Inicializar as penalidades de cada lobo
    alcateia[dim+3:, :] = 0
    
    #Inserir o lobo_1
    alcateia[:dim, 0] = parametros_rede["Lobo1"]
    
    return alcateia

In [10]:
'''
Teste de inicializar_alcateia()
'''
alcateia = inicializar_alcateia(12, rede, parametros_rede)
print('Primeiro lobo = {}'.format(alcateia[:,0].tolist()))
print('Tensões = {}'.format(alcateia[0:ng, 0]))
print('Taps = {}'.format(alcateia[ng:ng+nt, 0]))
print('Shunts = {}'.format(alcateia[ng+nt:ng+nt+ns, 0]))

Primeiro lobo = [1.045, 1.01, 1.07, 1.09, 0.978, 0.969, 0.9319999999999999, 0.19, inf, inf, 4.0, 0.0, 0.0, 0.0]
Tensões = [1.045 1.01  1.07  1.09 ]
Taps = [0.978 0.969 0.932]
Shunts = [0.19]


Resultado esperado:

    Primeiro lobo = [1.045, 1.01, 1.07, 1.09, 0.978, 0.969, 0.932, 0.19, inf, inf, 4.0, 0.0, 0.0, 0.0]
    Tensões = [1.045, 1.01, 1.07, 1.09]
    Taps = [0.978, 0.969, 0.932]
    Shunt = [0.19]

# Funções penalidade e função objetivo

* **Função Penalidade Senoidal para os taps dos transformadores do sistema**

A função penalidade senoidal para os taps do sistema (que pertencem a um conjunto discreto com passo constante) é dada pela equação:

\begin{equation}
    P_{tap} = \sum_{i = 1}^{n_t} \Bigl[sen \Bigl(\pi \dfrac{t_i}{tap_{step}}\Bigr)\Bigr] 
\end{equation}




* **Função Penalidade Senoidal para os shunts das barras do sistema**

A função penalidade senoidal para os shunts do sistema (que não possuem passo discreto constante) é dada pela equação

\begin{equation}
    P_{sen}(y) = \sum_{i = 1}^{n_y} \left[sen \left( \dfrac{y_i}{y_i^{sup} - y_i^{inf}}\pi +\alpha_i \right) \right]^2
    \label{pen-sen}
\end{equation}

Em que, $y_i^{sup}$ é o valor discreto superior mais próximo de $y_i$; $y_i^{inf}$ é o valor discreto inferior mais próximo de $y_i$; $\alpha_i$ é ($0 \leq \alpha_i \leq \pi$) é determinado de forma que $P_{sen}(y) = 0$ se, e somente se, as variáveis $y_i$, com $i = 1, 2, ..., n_y$, assumem valores discretos do conjunto $D_{y_i}$.

\begin{equation}
    \alpha_i = \left( \left \lceil \dfrac{y_i^{inf}}{y_i^{sup} - y_i^{inf}} \right \rceil - \dfrac{y_i^{inf}}{y_i^{sup} - y_i^{inf}} \right)\pi
    \label{ai}
\end{equation}

O operador $\lceil n \rceil$ representa o valor arredondado para cima do número $n$. Por construção, a função de penalidade senoidal é nula se $y_i$ assumir um valor discreto do intervalo $D_{y_i}$, para $i = 1, 2, ..., n_y$, e é positiva caso contrário. Caso a variável $y_i$ possua passo discreto constante, $\alpha_i$ também é constante.

* **Função Penalidade Estática para as tensões das barras do sistema**


A função de penalização estática para os limites das magnitudes de tensão das barras do sistema é

\begin{equation}
    P_V(V_{i }) = \sum_{i = 1}^{n_b} (V_i - V_i^{lim})^2
    \label{pen-v}
\end{equation}

Onde,
\begin{equation}
    V_i^{lim} =
    \begin{cases}
    \overline{V_i}, & \text{caso} \quad V_i > \overline{V_i}\\
    V_i, & \text{caso} \quad \underline{V_i} < V_i < \overline{V_i}\\
    \underline{V_i}, & \text{caso} \quad V_i < \underline{V_i}
    \end{cases}
\end{equation}

* **Função Objetivo**

A função objetivo do problema de FPOR (que deve ser minimizada) é:

\begin{equation*}
    f_{obj}(V, \theta, t) = \sum_{k,m \in \Omega}g_{km}\left(\dfrac{1}{t_{km}^2}V_k^2 + V_m^2 - 2\dfrac{1}{t_{km}}V_k V_m cos \theta_{km} \right)
\end{equation*}


In [11]:
def penalidade_senoidal_tap(alcateia = None, DEBUG = False, vec_debug = None):
    '''
    Esta função retorna a penalidade senoidal sobre os taps dos transformadores para toda a alcateia.
    Dado um tap t e um passo discreto s, a função penalidade senoidal é dada por:
        pen_sen_tap = sum {sen^2(t*pi/s)}    
    
    Inputs:
        -> alcateia
        
    Outputs:
        -> pen_taps: um vetor cuja forma é (nt, n_lobos) contendo as penalizações referentes aos taps para toda a alcateia
    '''
    if DEBUG:
        taps = vec_debug
        pen_taps = np.zeros(shape = vec_debug.shape)
    else:
        #Variável para receber os taps de todos os lobos
        taps = alcateia[ng:ng+nt, :]
        #Variável para armazenar as penalidades referentes aos taps de todos os lobos
        pen_taps = np.array(np.zeros((1, alcateia.shape[1])))
    
    #Executa a equação da penalização sem efetuar a soma
    taps = np.square(np.sin(taps*np.pi/tap_step))
    pen_taps = taps
    
    if not DEBUG:
        #Executa a soma ao longo das colunas da variável taps
        pen_taps = np.sum(taps, axis=0)
    
    threshold = np.less_equal(pen_taps, 1e-15)
    pen_taps[threshold] = 0.0

    #Deleta a variável taps
    del taps
    
    return pen_taps

In [12]:
'''
Teste de penalidade_senoidal_tap()
'''
print('Tap_step = {}'.format(tap_step))
print('Taps permitidos : {}'.format(parametros_rede['Valores_taps']))
print('\n')

vec_teste = np.array([0.9, 1.025, 1.1, 0.903, 1.014, 1.0666])
penalidade = penalidade_senoidal_tap(DEBUG = True, vec_debug = vec_teste)

print('Vetor_teste = {}'.format(vec_teste.tolist()))
print('Penalidade = {}'.format(penalidade.tolist()))

Tap_step = 0.00625
Taps permitidos : [0.9     0.90625 0.9125  0.91875 0.925   0.93125 0.9375  0.94375 0.95
 0.95625 0.9625  0.96875 0.975   0.98125 0.9875  0.99375 1.      1.00625
 1.0125  1.01875 1.025   1.03125 1.0375  1.04375 1.05    1.05625 1.0625
 1.06875 1.075   1.08125 1.0875  1.09375 1.1    ]


Vetor_teste = [0.9, 1.025, 1.1, 0.903, 1.014, 1.0666]
Penalidade = [0.0, 0.0, 0.0, 0.996057350657232, 0.4686047402352838, 0.7784378082441882]


Resultado esperado:

    Penalidade = [0.0, 0.0, 0.0, 0.9960573507, 0.4686047402, 0.7784378082]

In [13]:
def penalidade_senoidal_shunt(conjunto_shunts, alcateia = None, DEBUG = False, vec_debug = None):
    '''
    Esta função retorna a penalidade senoidal sobre os shunts de toda a 
    alcateia.
    
    Seja 'conjunto' a lista de valores discretos que o shunt admite: 
        conjunto = [b_1, b_2, b_3, b_4].
    Seja b um shunt
    
    A função pen_sen_shunt deve ser nula caso 'b' pertença a 'conjunto' e 
    maior que 0 caso contrário.
    
    Define-se a variável a função pen_sen_shunt para o caso de um único shunt b como:
        
        pen_sen_shunt = sen[ pi * (b /(b_sup - b_inf)) + alfa ]
    
    Onde:
        - b_sup: é o valor discreto superior mais próximo de 'b'
        - b_inf: é o valor discreto inferior mais próximo de 'b'
        - alfa: é uma variável escolhida para que pen_sen_shunt = 0 caso 'b' pertença a 'conjunto'
    
    Alfa é dada por:
        
        alfa = pi*[ ceil{b_inf/(b_sup - b_inf)} - b_inf/(b_sup - b_inf)]
        
        *ceil(x) é o valor de x arredondado para o inteiro superior mais próximo
    
    Inputs:
        -> alcateia 
        -> conjunto_shunts = conjunto de valores que cada shunt de 'alcateia' pode admitir
        
    Outputs:
        -> pen_shunts: um vetor cuja forma é (ns, n_lobos) contendo as penalizações referentes aos shunts para toda a alcateia
    
    '''
    if DEBUG:
        shunts = vec_debug
        alfa = np.zeros(shape=vec_debug.shape)
    else:
        #Variável para receber os shunts da alcateia
        shunts = alcateia[ng+nt:ng+nt+ns, :]
        #A variável alfa será um vetor no formato (ns, 1)
        alfa = np.zeros(shape=(ns, alcateia.shape[-1]))
    
    
    
    #Variáveis temporárias para armazenar b_inf's e b_sup's, obtidos pelas funções auxiliares descritas no ínicio do código
    shunts_sup = discreto_superior(shunts, conjunto_shunts)
    shunts_inf = discreto_inferior(shunts, conjunto_shunts)
    
    d = shunts_sup - shunts_inf
    if DEBUG:
        print('d: {}'.format(d))
    alfa = np.pi * (np.ceil(shunts_inf/d) - shunts_inf/d)
    
    pen_shunts = np.sin(alfa + np.pi*(shunts/d))
    pen_shunts = np.square(pen_shunts)
    if not DEBUG:
        pen_shunts = np.sum(pen_shunts, axis = 0)
    threshold = np.less_equal(pen_shunts, 1e-12)
    pen_shunts[threshold] = 0.0
    
    return pen_shunts

In [14]:
'''
Teste de penalidade_senoidal_shunt()
'''
v_teste= np.array([0.0, 0.19, 0.34,  0.39, 0.15, 0.21, 0.35])
p = penalidade_senoidal_shunt(conjunto_shunts[0], DEBUG = True, vec_debug = v_teste)
print('V_teste = {}'.format(v_teste))
print('Penalidade = {}'.format(p.tolist()))

d: [0.19       0.15       0.15       0.04999998 0.19       0.15
 0.04999998]
V_teste = [0.   0.19 0.34 0.39 0.15 0.21 0.35]
Penalidade = [0.0, 0.0, 0.0, 0.0, 0.377257226259472, 0.16543461458424602, 0.3454912896838885]


Resultado esperado:

    Penalidade = [0.0, 0.0, 0.0, 0.0, 0.37725726, 0.16543469, 0.34549150]

In [34]:
def funcao_objetivo_e_pen_v(rede, matriz_G, v_lim_sup, v_lim_inf):
    """
    E função  calcula a função objetivo para o problema de FPOR e também calcula
    a penalização das tensões que ultrapassam o limite superior ou ficam abaixo do limite
    inferior para cada agente de busca.
    
    Esta função não é vetorizada para toda a alcateia
    
    A função objetivo deste problema dá as perdas de potência ativa no SEP.
    
    f = sum g_km * [v_k^2 + v_m^2 - 2*v_k*v_m*cos(theta_k - theta_m)]
    
    A penalização das tensões é:
        
    pen_v = sum(v - v_lim)^2
        v_lim = v_lim_sup, se v > v_lim_sup
        v_lim = v, se v_lim_inf < v < v_lim_sup
        v_lim = v_lim_inf, se v < v_lim_inf
        
    Inputs:
        -> rede = sistema elétrico de testes
        -> matriz_G = matriz de condutância nodal do sistema
        -> v_lim_sup = vetor contendo os limites superiores de tensão nas barras do sistema
        -> v_lim_inf = vetor contendo os limites inferiores de tensão nas barras do sistema
    Outputs:
        -> f = função objetivo do problema de FPOR
        -> pen_v = penalidade de violação dos limites de tensão das barras
    
    """
    
    v_k = np.array([rede.res_bus.vm_pu.to_numpy(dtype = np.float64)])
    v_m = v_k.T
    
    theta_k = np.radians(np.array([rede.res_bus.va_degree.to_numpy(dtype = np.float64)]))
    theta_m = theta_k.T
    
    #Calculo da função objetivo
    f = np.power(v_k,2) + np.power(v_m,2) - 2*np.multiply(np.multiply(v_k,v_m),np.cos(theta_k-theta_m))
    f = np.multiply(matriz_G, f)
    f=np.squeeze(np.sum(np.array([f[np.nonzero(f)]])))
    
    #Calculo da penalidade das tensões
    
    #Violação do limite superior (v - v_lim), v > v_lim_sup
    v_up = v_k - v_lim_sup
    v_up = v_up[np.greater(v_up, 0.0)]
    
    #Violação do limite inferior (v-v_lim), v < v_lim_inf
    v_down = v_k - v_lim_inf
    v_down = v_down[np.less(v_down, 0.0)]
    
    pen_v = np.squeeze(np.sum(np.square(v_up)) + np.sum(np.square(v_down)))

    return f, pen_v

In [35]:
'''
Teste de funcao_objetivo_e_pen_v()
'''

matriz_G = parametros_rede["matriz_G"]
v_lim_sup = rede.bus.max_vm_pu.to_numpy(dtype = np.float64)
v_lim_inf = rede.bus.min_vm_pu.to_numpy(dtype = np.float64)
pp.runpp(rede, algorithm='fdbx')

func_obj, penalidade = funcao_objetivo_e_pen_v(rede, matriz_G, v_lim_sup, v_lim_inf)

print('Função objetivo = {}'.format(func_obj))
print('Penalidade = {}'.format(penalidade))

Função objetivo = 0.13393272357766134
Penalidade = 0.0010023089789910904


Resultado esperado:

    Função objetivo = 0.13393272357766134
    Penalidade = 0.0010023089789910904